Grenades and projectiles can have shrapnel and embed, all carbons can suffer embeds, some bullets can ricochet, sizable refactor of embedding

Ryll-Ryll opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
It annoyed me that we have a perfectly good frag grenade item, and a perfectly good shrapnel component, but no crossover episode between the two. This remedies that, and does a lot, lot more.

[![dreamseeker_2020-03-30_05-01-13.png](https://i.imgur.com/Rj8pHqNl.jpg)](https://i.imgur.com/Rj8pHqN.png)

[![dreamseeker_2020-03-30_05-01-26.png](https://i.imgur.com/nU5nYJxl.jpg)](https://i.imgur.com/nU5nYJx.png)

Big points:
- Adds new component: pellet_cloud, which can be used by ammo casings, guns, and landmines to spray shrapnel and display aggregate hit messages ("You're hit by 6 buckshot pellets!" vs "You're hit by the buckshot pellet in the X" x6). All gun ammo that shoot multiple pellets now use this component on firing.
- Adds stingbangs, premium less-lethal grenades that shoot off lots of stinger pellets, to cargo. Frag grenades are also reworked to have smaller booms, but shoot off lots of shrapnel shards. You can jump on top of these grenades to absorb a portion of the shrapnel to save those around you! There's an achievement for dying this way, called "Look Out, Sir!"
- Projectiles can now embed items/shrapnel. Adds .38 DumDum ammo to cargo that does less damage and has negative armor pen, but can embed in people. This is the only ammo that currently embeds.
- Bullets can now ricochet off walls, structures, and machinery (harder surfaces are more likely to ricochet). Only standard .38 and Match Grade .38/.357/L6 ammo can ricochet, with Match Grade being much better at ricocheting. You can buy Match Grade .38 from cargo and Match Grade L6 ammo from the nuke uplink, while Match .357 is admin only.
- Armor now protects you from harmful embeds, taking the better of the bullet/bomb armor on the affected limb. Armor penetration can modify this of course, and many blunt embeds like stingbangs and DumDum bullets are significantly worse if you have even 1 armor. 

Other misc fixes/changes
- Refactored the embed element a bunch and fixed it creating new elements for every instance rather than expected bespoke behavior. There are new /obj/item helpers for modifying and adding embedding.
- Fixes #49989: Spears can no longer embed in turfs cause their sprite is annoying to me, it's generally harder for most things to embed in turfs
- Fixes #49741: New carbon helpers for removing embedded objects
- Fixes #46416: Handles embedded objects getting qdel'd or moved while embedded
- Renamed the old shrapnel component for RPG loot to MIRV to avoid confusion
- Repathed frag grenades from under minibombs to under base grenades, and added explosion vars to base grenades

<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Fixes a bunch of janky design with embeds, adds lots of new avenues for projectile and grenade variety, ricochets and collateral damage are fun! 

<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl: Ryll/Shaps
add: Adds stingbangs to cargo (and one in the sec vendor premium), premium less-lethal grenades that shoot off a giant swarm of stingball pellets to help incapacitate swarms of people in tight quarters. You can jump on top of a live one to be a hero and absorb a bunch of shrapnel, same with  frag grenades. There's even an achievement for dying to a grenade you jumped on!
add: Projectiles can now embed in people! Or at least grenade shrapnel and the new .38 DumDum ammo, now available in cargo, can. DumDum rounds excel against unarmored targets, but are pricey and do poorly against armored targets.
add: Bullets can now ricochet! Or at least, standard .38 and the new .38/L6 Match Grade ammo can. Match Grade ammo is finely tuned to ricochet easier and seek targets off bounces better, and can be purchased from cargo (for the .38) or nuke ops uplink (for the L6), but standard .38 ammo has a chance to ricochet as well.
tweak: Frag grenades now have smaller explosions but shoot off a bunch of devastating shrapnel, excellent for soft targets!
tweak: Shotguns and other multi-pellet guns now print aggregate messages, so you'll get one "You've been hit by 6 buckshot pellets!" rather than 6 "You've been hit by the buckshot pellet in the X!" messages. Bye bye lag!
balance: Armor can now protect against embedding weapons, taking the best of either the bullet or bomb armor for the limb in question away from the embed chance. Some weapons are better at piercing armor than others!
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

trollbreeder commented:
This will make remove embed surgery a lot more useful.

TiviPlus commented:
FRAG OUT
GET DOWN MR. CLOWN



Good PR please finish

InsaneKatrina commented:
If this is the point where bullets start getting stuck into every part of a person's body then I say _yes._

ArcaneDefence commented:
Oh god. Even more walls of text from a shotgun loaded with buckshot being spammed at people

Ryll-Ryll commented:
I can see about silencing some of the messages or doing aggregate messages for "swarm" guns like buckshot, or maybe just keep the embeds for slug/bullet weapons failing that

Whoneedspacee commented:
based

Bawhoppen commented:
Bawhoppen did not know that fragmentation grenades released fragments when Bawhoppen added them to the game originally. Bawhoppen based their knowledge of military technology off of Half Life 1, where fragmentation grenades simply made a medium-sized explosion.

Ryll-Ryll commented:
fair.......... but.... the name.........

Bawhoppen commented:
Bawhoppen is not disagreeing that they should do this also... just explaining why they didn't try to add it originally (also knowing how to code it was another significant reason).

Tlaltecuhtli commented:
can you make embedding support organ damage? so we can have bullet shrapnells slowly  kill you from inside instead of being ninja stars but small

Ryll-Ryll committed:
Basic helpers first


functional


late night drafts ftw please don't review this, it works but it sucks

Ryll-Ryll commented:
This is still alive, I'm working on making pellet clouds (anything that works like a shotgun, including grenades with shrapnel) where multiple hits on a target will only show one hit with how many pellets hit them, but I'd like some feedback if people think it'd be useful or not. Also figuring out how to mesh this with the shrapnel component
[![dreamseeker_2020-03-10_00-19-32.png](https://i.imgur.com/UHHpsV2.png)](https://i.imgur.com/UHHpsV2.png)

Rohesie commented on code/datums/components/pellet_cloud.dm:
```diff
@@ -0,0 +1,121 @@
+/datum/component/pellet_cloud/Destroy(force, silent)
+	if(holdup)
+		return QDEL_HINT_LETMELIVE
+	. = ..()
```
```suggestion
	return ..()
```

Rohesie commented on code/datums/components/pellet_cloud.dm:
```diff
@@ -0,0 +1,121 @@
+
+/datum/component/pellet_cloud/proc/create_pellets(obj/item/ammo_casing/A, target, user, fired_from)
+	aim_target = target
+	for(var/pelleties = 0, pelleties < num_pellets, pelleties++)
```
```suggestion
	for(var/i in 1 to num_pellets)
```
You'll want the C loop for cases where you want to manipulate the control variable and have it affect how many times the proc is run.
If you don't, the list loop skips extra checks and just does what is intended.

Rohesie commented on code/datums/components/pellet_cloud.dm:
```diff
@@ -0,0 +1,121 @@
+
+		pellets += P
+		RegisterSignal(P, COMSIG_PROJECTILE_SELF_ON_HIT, .proc/pellet_hit)
+		RegisterSignal(P, COMSIG_PROJECTILE_SWEET_FA, .proc/pellet_range)
```
You'll want to register those signals before firing it.

Rohesie commented on code/datums/components/pellet_cloud.dm:
```diff
@@ -0,0 +1,121 @@
+	var/list/all_the_turfs_were_gonna_lacerate = RANGE_TURFS(radius, target) - RANGE_TURFS(radius-1, target)
+	num_pellets = all_the_turfs_were_gonna_lacerate.len
+	var/turf/target_turf = get_turf(target)
+	for(var/turf/shootat_turf in all_the_turfs_were_gonna_lacerate)
```
```suggestion
	for(var/t in all_the_turfs_were_gonna_lacerate)
		var/turf/shootat_turf = t
```
RANGE_TURFS returns turf type only already.

Rohesie commented on code/datums/components/pellet_cloud.dm:
```diff
@@ -0,0 +1,121 @@
+		RegisterSignal(P, COMSIG_PROJECTILE_SELF_ON_HIT, .proc/pellet_hit)
+		RegisterSignal(P, COMSIG_PROJECTILE_SWEET_FA, .proc/pellet_range)
+		P.fire()
+		pellets += P
```
Couldn't `fire()` potentially make the projectile hit a target and get it deleted before running the addition of the object to the list?

Rohesie commented on code/datums/components/pellet_cloud.dm:
```diff
@@ -0,0 +1,121 @@
+	if(!targets_hit[target])
+		targets_hit[target] = 1
+	else
+		targets_hit[target] += 1
```
You can also simply do this:
```suggestion
	targets_hit[target]++
```

Rohesie commented on code/datums/components/pellet_cloud.dm:
```diff
@@ -0,0 +1,121 @@
+/datum/component/pellet_cloud/proc/finalize()
+	var/i = 0
+	for(var/atom/targeties in targets_hit)
+		i++
```
What is `i` doing here?

Rohesie commented on code/datums/components/pellet_cloud.dm:
```diff
@@ -0,0 +1,113 @@
+/datum/component/pellet_cloud/Destroy(force, silent)
+	if(holdup)
+		return QDEL_HINT_LETMELIVE
+	QDEL_NULL(pellets)
```
QDeling a list?

Rohesie commented on code/datums/components/pellet_cloud.dm:
```diff
@@ -0,0 +1,113 @@
+/datum/component/pellet_cloud
+	var/projectile_type
+	var/num_pellets
+	var/list/pellets = list()
```
Lazy list perhaps?

Rohesie commented on code/datums/components/shrapnel.dm:
```diff
@@ -4,30 +4,41 @@
+
+///datum/component/shrapnel/proc/do_shrapnel(mob/firer, atom/target)
 
+*/
```
Could you clean these leftover comments?

Rohesie commented on code/datums/elements/embed.dm:
```diff
@@ -129,3 +137,23 @@
+
+	Detach(P)
+	//QDEL_NULL(P)
+	return
```
WIP?

Ryll-Ryll commented:
Whole bunch more changes along the way, possibly bullet ricochets (while I was working on the stingbang grenades I thought it might be neat for the .38) and neatening up of embedding code. Also renaming the shrapnel component to mirv since it functions differently than this (it's on a per-gun basis rather than per-munition)

Rohesie commented:
Might want to consider atomizing if it starts getting too big.

Ryll-Ryll commented:
hmmmmm maybe you have a point

Ryll-Ryll committed:
better swarm


just had another idea i need to flit to


another update


oh boy


item proc


Basic helpers first


functional


late night drafts ftw please don't review this, it works but it sucks


just had another idea i need to flit to


another update


oh boy

Ryll-Ryll committed:
i should really stop trying to figure out git issues while xcomming

Ryll-Ryll committed:
update embed procs and helpers and such

Ryll-Ryll commented:
I'll write up a full thing in the next few days, but the broad strokes as I wrap this up are:

- Fixed prior shoddiness in embed element that created new elements for every object rather than expected bespoke grouping.
- Ammo casings that have more than one pellet will now only display aggregate hit counts: Instead of getting five messages about getting hit by buckshot pellets, the game will wait until all pellets have hit something or disappeared, and then print one visible_message per target. Same for grenades that have shrapnel..
- Projectiles can now have "payload" items that spawn on impact and try to embed in the target. I added .38 DumDum rounds that can embed to cargo for 1200cr/speedloader to show this off.
- Grenades can now shoot off projectiles as shrapnel, which combined with the above, make for the desired frag grenades. Frag grenades now shoot off shrapnel shards, and added less-lethal "stingbangs" to cargo (1400cr/1, 2500cr/5) which are flashbangs with a smaller stun bang, but shoot off a ton of ricocheting stingballs.
- Continuing on the above, ballistics can now ricochet off hard surfaces if set to be able to. By default, only the standard .38 ammo can ricochet, but this also adds Match Grade ammo for the .38 (1200cr/speedloader in cargo), L6 SAW (10TC/box mag), and .357 (admin only). Frag shrapnel and stingballs can also bounce. Walls and especially reinforced walls are easier to bounce bullets off, but windows, airlocks, and other structures have chances to as well.
- Renamed old shrapnel component to mirv, to help differentiate it from pellet clouds (mirv component have no aggregate messages and attaches to guns during RPG loot to modify anything it fires, pellet clouds attach to individual munitions like ammo casings and grenades and use aggregate messages)

Ryll-Ryll committed:
new ammos, new nades, new features

Ryll-Ryll committed:
test

trollbreeder commented:
![image](https://user-images.githubusercontent.com/25937686/77831795-8cee8700-7131-11ea-95f0-1530db333dd3.png)
Nice.

LordSaladin commented:
what if we had BBs  and BB guns.

Ryll-Ryll committed:
achievement and some item changes and mines

Ryll-Ryll committed:
just before i remove the testing's in case I need them later!

Ryll-Ryll committed:
document and hold off on prying embeds from people

Ryll-Ryll committed:
finishing touches

Ryll-Ryll commented:
FINALLY

I'm calling it here, feature complete. Review at will, please

Rohesie commented on code/datums/components/embedded.dm:
```diff
@@ -64,11 +66,14 @@
 		return COMPONENT_INCOMPATIBLE
 
+	if(part)
+		L = part
```
Could you rename `L` to `limb` or something more readable?

Rohesie commented on code/datums/components/embedded.dm:
```diff
@@ -124,26 +131,30 @@
-	var/mob/living/carbon/human/victim = parent
-	L = pick(victim.bodyparts)
+	var/mob/living/carbon/victim = parent
+	if(!L || !istype(L))
```
```suggestion
	if(!istype(L))
```
`istype(falsey value)` will return false already.

Rohesie commented on code/datums/components/embedded.dm:
```diff
@@ -157,27 +168,33 @@
 	if(do_after(victim, time_taken, target = victim))
 		if(!weapon || !L || weapon.loc != victim || !(weapon in L.embedded_objects))
-			qdel(src)
+			QDEL_NULL(src)
```
`src = null` is not really kosher

Rohesie commented on code/datums/components/pellet_cloud.dm:
```diff
@@ -0,0 +1,249 @@
+	if(magnitude < 1)
+		CRASH("Invalid magnitude [magnitude] < 1 on pellet_cloud, parent: [parent]")
+		qdel(src, TRUE)
+		return
```
`CRASH` acts like a `return`
The code under it won't execute.
Use `stack_trace` instead.

Rohesie commented on code/datums/components/pellet_cloud.dm:
```diff
@@ -0,0 +1,249 @@
+			else //Smart spread
+				spread = round((i / num_pellets - 0.5) * distro)
+
+		RegisterSignal(A.BB, COMSIG_PROJECTILE_SELF_ON_HIT, .proc/pellet_hit)
```
Could you use more clear variables? `A.BB` is not very informative.

Ryll-Ryll replied:
BB is the name ammo casings use for their projectile (presumably BB like a metal ball bearing in a BB gun), would a comment clarifying this suffice since I'm just using vars that were already there

Rohesie replied:
At least fix the `A` then. These bad standards proliferate.
`BB` is not kosher either but it's not your fault.

Rohesie commented on code/datums/components/pellet_cloud.dm:
```diff
@@ -0,0 +1,249 @@
+		RegisterSignal(A.BB, list(COMSIG_PROJECTILE_RANGE_OUT, COMSIG_PARENT_QDELETING), .proc/pellet_range)
+		pellets += A.BB
+		if(!A.throw_proj(target, targloc, shooter, params, spread))
+			return 0
```
```suggestion
			return
```

Rohesie commented on code/datums/components/pellet_cloud.dm:
```diff
@@ -0,0 +1,249 @@
+		else if(radius >= 2)
+			L.visible_message("<b><span class='danger'>[L] heroically covers \the [parent] with [L.p_their()] body, absorbing some of the shrapnel!</span></b>", "<span class='userdanger'>You heroically cover \the [parent] with your body, absorbing some of the shrapnel!</span>")
+			magnitude_absorbed += 2
+		else
```
You can use a `switch` statement here:
```cpp
switch(radius)
    if(5 to INFINITY)
    if(2 to 4)
    else
```
Is the visible message different on each case here? They look the same.
If so, you could put it before or after the `switch`

Rohesie commented on code/datums/components/pellet_cloud.dm:
```diff
@@ -0,0 +1,249 @@
+			magnitude_absorbed = radius
+
+		var/pellets_absorbed
+		pellets_absorbed = (radius ** 2) - ((radius - magnitude_absorbed - 1) ** 2)
```
```suggestion
		var/pellets_absorbed = (radius ** 2) - ((radius - magnitude_absorbed - 1) ** 2)
```

Rohesie replied:
Alternatively just do `. = (radius ** 2) - ((radius - magnitude_absorbed - 1) ** 2)`

Rohesie commented on code/datums/components/pellet_cloud.dm:
```diff
@@ -0,0 +1,249 @@
+	var/total_pellets_absorbed
+
+	for(var/M in martyrs)
+		var/mob/living/L = M
```
I never know if `L` means `living` or `limb`
Could you use something like `living_martyrs` or something like that?

Rohesie commented on code/datums/components/pellet_cloud.dm:
```diff
@@ -0,0 +1,249 @@
+/datum/component/pellet_cloud/proc/finalize()
+	var/obj/projectile/P = new projectile_type
+	var/proj_name = initial(P.name)
+	QDEL_NULL(P)
```
No need to create and delete a projectile to get its name. You can do it like this:
```cpp
var/obj/projectile/proj
var/proj_name = initial(proj.name)
```

Rohesie commented on code/datums/components/pellet_cloud.dm:
```diff
@@ -0,0 +1,249 @@
+			to_chat(target, "<span class='userdanger'>You're hit by [num_hits] [proj_name]s!</span>")
+
+	if(purple_hearts)
+		for(var/M in purple_hearts)
```
```suggestion
	for(var/M in purple_hearts)
```
No need for the `if`. If there's none, the `for` will simply not trigger.

Rohesie commented on code/datums/components/pellet_cloud.dm:
```diff
@@ -0,0 +1,249 @@
+
+/datum/component/pellet_cloud/Destroy(force, silent)
+	if(not_done_yet)
+		return QDEL_HINT_LETMELIVE
```
Is this necessary? Can't you have it wrap up things in `Destroy()`?

Rohesie commented on code/datums/elements/embed.dm:
```diff
@@ -1,22 +1,20 @@
 /datum/element/embed
 	element_flags = ELEMENT_BESPOKE
 	id_arg_index = 2
+	var/initialized = FALSE /// whether we can skip assigning all the vars (since these are bespoke elements, we don't have to reset the vars every time we attach to something, we already know what we are!)
```
This is generic behavior, should be added to `/datum/element`

Rohesie commented on code/datums/elements/embed.dm:
```diff
@@ -30,102 +28,196 @@
+		RegisterSignal(target, COMSIG_PROJECTILE_SELF_ON_HIT, .proc/checkEmbedProjectile)
 
 
 /datum/element/embed/Detach(obj/item/target)
```
```suggestion
/datum/element/embed/Detach(obj/target)
```

Ryll-Ryll replied:
Oh, very good catch

Ryll-Ryll committed:
1/2 review

Rohesie commented on code/datums/components/embedded.dm:
```diff
@@ -124,81 +131,105 @@
-/datum/component/embedded/proc/ripOutHuman()
-	var/mob/living/carbon/human/victim = parent
+/datum/component/embedded/proc/ripOutHuman(datum/source, obj/item/I, obj/item/bodypart/limb)
+	if(I != weapon || limb != limb)
```
```suggestion
	if(I != weapon || src.limb != limb)
```
When a local variable has the same name than an input variable, that's when you use the `src.` prefix. Else you are comparing the variable to itself.
If you mean to reference the `obj/item/bodypart/limb` that's being input in the proc, use `limb`.
If you want to reference the `var/obj/item/bodypart/limb` defined for the component, use `src.limb`.

Ryll-Ryll replied:
woopsie, the chaos of mass replace

Rohesie commented on code/datums/components/embedded.dm:
```diff
@@ -124,81 +131,105 @@
+		if(!victim.has_embedded_objects())
+			victim.clear_alert("embeddedobject")
+			SEND_SIGNAL(victim, COMSIG_CLEAR_MOOD_EVENT, "embedded")
+		QDEL_NULL(src)
```
Replace every instance of `QDEL_NULL(src)` with `qdel(src)`
There's no need to null `src`.

Rohesie commented on code/datums/components/pellet_cloud.dm:
```diff
@@ -0,0 +1,237 @@
+
+	if(magnitude < 1)
+		stack_trace("Invalid magnitude [magnitude] < 1 on pellet_cloud, parent: [parent]")
+		qdel(src, TRUE)
```
Do you need to force the deletion?

Ryll-Ryll replied:
I guess I can just give it a magnitude of 1 while leaving the stack trace

Rohesie commented on code/datums/components/pellet_cloud.dm:
```diff
@@ -0,0 +1,237 @@
+			martyr.client.give_award(/datum/award/achievement/misc/lookoutsir, martyr)
+
+	qdel(parent)
+	qdel(src)
```
Doesn't the parent deletion already handle the component one?

Rohesie commented on code/game/atoms.dm:
```diff
@@ -16,6 +16,13 @@
+	var/flags_ricochet = NONE
+
+	///When a projectile tries to ricochet off this atom, the projectile ricochet chance is multiplied by this
+	var/ricochet_chance_mod = 1.0
```
```suggestion
	var/ricochet_chance_mod = 1
```
No need for extra zeros, they make mass-replacements harder.

Rohesie commented on code/game/objects/items.dm:
```diff
@@ -207,18 +207,16 @@ GLOBAL_VAR_INIT(embedpocalypse, FALSE) // if true, all items will be able to emb
-	if(embedding)
-		AddElement(/datum/element/embed, embedding)
-	else if(GLOB.embedpocalypse)
+	// this proc says it's for initializing components, but we're initializing elements too because it's you and me against the world >:)
```
```suggestion
```
Duplicate comment?

Ryll-Ryll replied:
oh woopsie

Rohesie commented on code/game/objects/items.dm:
```diff
@@ -207,18 +207,16 @@ GLOBAL_VAR_INIT(embedpocalypse, FALSE) // if true, all items will be able to emb
-		AddElement(/datum/element/embed, embedding)
 		name = "pointy [name]"
-	else if(GLOB.stickpocalypse)
+	else if(!LAZYLEN(embedding) && GLOB.stickpocalypse)
```
You can nest these checks under a `if(!LAZYLEN(embedding)` and avoid the duplicate check.

Rohesie commented on code/game/objects/items.dm:
```diff
@@ -945,3 +949,47 @@ GLOBAL_VAR_INIT(embedpocalypse, FALSE) // if true, all items will be able to emb
+
+///For when you want to add/update the embedding on an item. Uses the vars in [/obj/item/embedding], and defaults to config values for values that aren't set. Will automatically detach previous embed elements on this item.
+/obj/item/proc/updateEmbedding()
+	if(!LAZYLEN(embedding) || !islist(embedding))
```
```suggestion
	if(!islist(embedding) || !LAZYLEN(embedding))
```
Check if it's a list before using a list-specific check.

Rohesie commented on code/game/objects/items/grenades/grenade.dm:
```diff
@@ -18,6 +18,15 @@
+	// dealing with creating a [/datum/component/pellet_cloud] on prime
+	var/shrapnel_type /// if set, will spew out projectiles of this type
+	var/shrapnel_radius /// the higher this number, the more projectiles are created as shrapnel
+	var/shrapnel_initialized
```
Put the documentation on the line over the variable.

Rohesie commented on code/game/objects/items/grenades/grenade.dm:
```diff
@@ -80,12 +89,23 @@
+
+	SEND_SIGNAL(src, COMSIG_GRENADE_PRIME)
+	if(ex_dev || ex_heavy || ex_light || ex_flame)
+		explosion(src.loc,ex_dev,ex_heavy,ex_light,flame_range = ex_flame)
```
```suggestion
		explosion(loc, ex_dev, ex_heavy, ex_light, flame_range = ex_flame)
```

Rohesie commented on code/modules/mob/living/living_defense.dm:
```diff
@@ -1,7 +1,11 @@
+	if(armor > 0 && silent)
+		armor = max(0, armor - armour_penetration)
+		return armor
+
```
This proc is doing a lot of superfluous operations. It can be improved.

Rohesie commented on code/modules/projectiles/ammunition/_firing.dm:
```diff
@@ -1,17 +1,21 @@
-			newshot()
+	else
+		if(isnull(BB))
+			return 0
```
```suggestion
			return FALSE
```

Ryll-Ryll committed:
a bit more neatening up

Rohesie commented on code/modules/projectiles/projectile.dm:
```diff
@@ -43,9 +43,14 @@
+	var/ricochet_decay_damage = 0.7 /// 0-1 (or more, I guess) multiplier, the projectile's damage is modified by multiplying this after each ricochet
+	var/ricochet_auto_aim_range = 0 /// On ricochet, if nonzero, we consider all mobs within this range of our projectile at the time of ricochet to home in on like Revolver Ocelot, as governed by ricochet_auto_aim_angle
+	var/ricochet_auto_aim_angle = 30 /// On ricochet, if ricochet_auto_aim_range is nonzero, we'll consider any mobs within this range of the normal angle of incidence to home in on, higher = more auto aim
+	var/ricochet_incidence_leeway = 30 /// the angle of impact must be within this many degrees of the struck surface, set to 0 to allow any angle
```
Add the documentation comments on the line over each of the variables.

Rohesie commented on code/modules/surgery/bodyparts/helpers.dm:
```diff
@@ -137,29 +137,25 @@
-	var/turf/T = get_turf(src)
+///Remove a specific embedded item from the carbon mob
+/mob/living/carbon/proc/remove_embedded_object(obj/item/I)
+	SEND_SIGNAL(src, COMSIG_HUMAN_EMBED_REMOVAL, I)
```
In general we don't want signals that are disguised procs.
If you are calling `remove_embedded_object()` on many places you can make several signals, so we can better track their sources.

Rohesie commented on code/modules/projectiles/projectile.dm:
```diff
@@ -43,9 +43,14 @@
+	var/ricochet_decay_damage = 0.7 /// 0-1 (or more, I guess) multiplier, the projectile's damage is modified by multiplying this after each ricochet
+	var/ricochet_auto_aim_range = 0 /// On ricochet, if nonzero, we consider all mobs within this range of our projectile at the time of ricochet to home in on like Revolver Ocelot, as governed by ricochet_auto_aim_angle
+	var/ricochet_auto_aim_angle = 30 /// On ricochet, if ricochet_auto_aim_range is nonzero, we'll consider any mobs within this range of the normal angle of incidence to home in on, higher = more auto aim
+	var/ricochet_incidence_leeway = 40 /// the angle of impact must be within this many degrees of the struck surface, set to 0 to allow any angle
```
Put the documentation on top of each variable.

Ryll-Ryll commented:
i LIED now all carbons (IE monkeys and xenos) can be embedded into and can remove embeds from themselves same was as humans can, interacting with themselves and clicking the link

I make no claim to whether this is balanced or not but I was gonna go nuts if I didn't generalize the embed code cause there was no reason code-wise for it to be human only and not carbons in general

[![dreamseeker_2020-03-31_03-27-23.png](https://i.imgur.com/GQdbZiGl.jpg)](https://i.imgur.com/GQdbZiG.png)

[![dreamseeker_2020-03-31_03-30-39.png](https://i.imgur.com/JYiE0e5l.jpg)](https://i.imgur.com/JYiE0e5.png)

Ryll-Ryll committed:
fixes broken incompatible component error message

Ryll-Ryll committed:
ok so basically im monky
